src/core_plugins/timelion/public/panels/timechart/schema.js

```diff
@@ -85,6 +86,49 @@ export default function timechartFn(Private, config, $rootScope, timefilter, $co
           colors: ['#01A4A4', '#C66', '#D0D102', '#616161', '#00A1CB', '#32742C', '#F18D05', '#113F8C', '#61AE24', '#D70060']
         };
 
+        $scope.chart.forEach((series, seriesIndex) => {
+          if (!series.color) {
+            const colorIndex = seriesIndex % defaultOptions.colors.length;
+            series.color = defaultOptions.colors[colorIndex];
+          }
+        });
+
+        const HIGHLIGHT_NOT_SET = -1;
+        let hightlightedSeries = HIGHLIGHT_NOT_SET;
```

Comment:
This algorithm does a clean sweep with `restoreColors`, and then reapplies the colors. It uses a temp-property, `_originalColor` on the series object to keep track of older state. Is there a way we can simplify this, without the need for sweeping past the data twice and having to maintain that temp-property

 in-sync?

Would this work, using a Map? Maybe I'm forgetting something obvious.

[patch.txt](https://github.com/elastic/kibana/files/1528628/patch.txt)

This is the relevant part:
```
        const originalColorMap = new Map();
        $scope.chart.forEach((series, seriesIndex) => {
          if (!series.color) {
            const colorIndex = seriesIndex % defaultOptions.colors.length;
            series.color = defaultOptions.colors[colorIndex];
            originalColorMap.set(series, series.color);
          }
        });

        const HIGHLIGHT_NOT_SET = -1;
        let hightlightedSeries = HIGHLIGHT_NOT_SET;

        function unhighlightSeries() {
          if (hightlightedSeries === HIGHLIGHT_NOT_SET) {
            return;
          }
          hightlightedSeries = HIGHLIGHT_NOT_SET;
          $scope.chart.forEach((series, seriesIndex) => {
              series.color = originalColorMap.get(series); //just reset the colors
          });
          drawPlot($scope.chart);
        }
        
        $scope.highlightSeries = _.debounce(function (id) {
          if (hightlightedSeries === id) {
            return;
          }

          hightlightedSeries = id;
          $scope.chart.forEach((series, seriesIndex) => {
            if (seriesIndex !== id) {
              series.color = 'rgba(128,128,128,0.1)';//mark as grey
            } else {
              series.color = originalColorMap.get(series);//color it like it was
            }
          });
          drawPlot($scope.chart);
        }, DEBOUNCE_DELAY);
```